contracts/contracts/FleekPausable.sol

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.7;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+error ContractIsPaused();
+error ContractIsNotPaused();
+error ContractIsNotPausable();
+error PausableIsSetTo(bool state);
+
+abstract contract FleekPausable is Initializable {
+    /**
+     * @dev Emitted when the pause is triggered by `account` and set to `isPaused`.
+     */
+    event PausedChange(bool indexed isPaused, address account);
+
+    /**
+     * @dev Emitted when the pausable is triggered by `account` and set to `isPausable`.
+     */
+    event PausableChange(bool indexed isPausable, address account);
+
+    bool private _paused;
+    bool private _canPause; // TODO: how should we verify if the contract is pausable or not?
```

Comment:
We need to think about when the contract is not pausable first. For example, when a transaction is initiated before the pause, it will revert if the contract's status changes.
At the same time, we need to consider the possibility of bad actors finding a balance of enough gas to keep the transaction pending but not enough priority gas for it to be picked up faster. They can make the contract impossible to pause externally and that's a bad position to be in.

Reply:
Agreed, this needs to be discussed. cc @jsonsivar

Reply:
I'm not sure I understand the issue here yet. The _canPause is like a feature flag that only Fleek controller can call (and later can be set by a governance contract or multisig). The only time we will need to use this pause feature imo is if it's an emergency and we need to pause it to protect peoples assets. In that example like a vulnerability is discovered with our contracts or EVM e.g., I don't think it matters of any transactions initiated before the pause transaction fails right? I could be missing something here.